orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,301 @@
+			return idx, err
+		}
+	}
+	return 0, errors.New("Endpoint not in the list")
+}
+
+func (bf *BlockFetcher) shuffleEndpoint() error {
+	// Iterate through the bf.Endpoints and choose the next endpoint
+	// which is not suspected
+	// shuffle endpoint if there is one
+
+	currIdx, err := bf.getIdxForEndpoint(bf.currentEndpoint.Endpoint)
+	if err != nil {
+		return nil
+	}
+	// Lets us increase the current idx by 1, for shuffle
+	// shuffle here is just iterating to next endpoint
+	// increment is circular, i.e after last element, an increment will lead to the first element.
+
+	nextIdx := (currIdx + 1) % len(bf.Endpoints)
+	// check nextIdx in suspect list
+
+	for {
+		// TODO: keep a bound on this loop so that it doesn't go infinite
+		if !bf.suspects.has(bf.Endpoints[nextIdx].Endpoint) {
+			bf.Logger.Debugf("Endpoint Shuffled. Current Endpoint %s", bf.currentEndpoint.Endpoint)
+			bf.currentEndpoint = bf.Endpoints[nextIdx]
+			break
+		}
+		currIdx = nextIdx
+		nextIdx = (currIdx + 1) % len(bf.Endpoints)
+	}
+
+	bf.LastShuffledAt = bf.TimeNow()
+	return nil
+}
+
+func (bf *BlockFetcher) isShuffleTime() bool {
+	if bf.ShuffleTimeout == time.Duration(0) {
+		// It should return only false if Time based Shuffling is not enabled
+		bf.Logger.Debugf("Shuffle Disabled.")
+		return false
+	}
+
+	shuffle := time.Now().After(bf.LastShuffledAt.Add(bf.ShuffleTimeout))
+	bf.Logger.Debugf("Shuffle Enabled. Time to shuffle %v", shuffle)
+	return shuffle
+}
+
+func (bf *BlockFetcher) isEndpointByzantine(ec EndpointCriteria) bool {
+	// probe the other endpoints
+	bf.Logger.Infof("Checking for BFT Behavior of %s", ec.Endpoint)
+	// choose (MaxByzantineFaults + 1) endpoints other than currentEndpoint
+	//currIdx, err := bf.getIdxForEndpoint(bf.currentEndpoint.Endpoint)
+	currIdx, _ := bf.getIdxForEndpoint(ec.Endpoint)
+	// TODO: handle above error case
+
+	endpointsToProbe := []EndpointCriteria{}
+	for i := int64(currIdx + 1); i <= (bf.MaxByzantineFaults + int64(currIdx)); i++ {
+		// add the next (bf.MaxByzantineFaults + 1)
+		endpointsToProbe = append(endpointsToProbe, bf.Endpoints[i%(int64(len(bf.Endpoints)))])
+	}
+
+	// add a Sync.WaitGroup and probe using go routines
+	waitGrp := sync.WaitGroup{}
+	// recieve the blocks using channels to a slice
+
+	// verify the slice and decide whether current node had byzantine behavior
+	type bps struct {
+		block *common.Block
+		ec    EndpointCriteria
+	}
+
+	bpsInfo := make(chan *bps, len(endpointsToProbe))
+
+	waitGrp.Add(len(endpointsToProbe))
+	// start probing
+	for _, ec := range endpointsToProbe {
+		// get a puller
+
+		go func(ec EndpointCriteria) {
+			defer waitGrp.Done()
+			config := bf.Config
+			config.Endpoints = []EndpointCriteria{ec}
+			puller := bf.BlockSourceFactory(config)
+			// To prevent connection leak
+			defer puller.Close()
+
+			// set endpoints
+			puller.UpdateEndpoints([]EndpointCriteria{ec})
+
+			// pull block
+			// TODO: Later pull attestations at this point
```

Comment:
we merged the attestations, so you can now use them.

Reply:
sure